AWS shared VPC install

yunjiang29 opened the pull request:
New workflows:
```
cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts
cucushift-installer-rehearse-aws-ipi-shared-vpc-phz
cucushift-installer-rehearse-aws-ipi-shared-vpc-phz-sts
```

New jobs:
```
periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-tp-p2-f14
periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p2-f14
periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-tp-p2-f14
```



TODO (maybe in separate PR):
- [ ] Narrow down permissions, see comment https://github.com/openshift/release/pull/39946#discussion_r1228261061

yunjiang29 commented:
/wip

yunjiang29 commented:
/hold

yunjiang29 commented:
@jiajliu @gpei please review this PR when you are available, this PR is still in draft, and hasn't been tested yet.  Thanks.

barbacbd commented:
/cc @barbacbd

patrickdillon commented on ci-operator/step-registry/aws/deprovision/route53/private-hosted-zone/aws-deprovision-route53-private-hosted-zone-commands.sh:
```diff
@@ -6,7 +6,12 @@ set -o pipefail
 
-export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred"
+if [[ ${ENABLE_SHARED_PHZ} == "yes" ]]; then
+  export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred_shared_account"
```
What are the contents of `.awscred_shared_account`? Is it the credentials for the other CI account; so that if you're using the `openshift-ci` account, `awscred_shared_account` contains the credentials for  `openshift-ci-2`?

yunjiang29 replied:
Right, Account A(VPC and PHZ owner)'s credential is saved in  `.awscred_shared_account`:
```
[default]
aws_access_key_id = ACCOUNT-A-KEY
aws_secret_access_key = ACCOUNT-A-SEC
aws_account_id = ACCOUNT-A-ID
```

Another way to handle this is to save all credentials in one `.awscred` file, like:
```
[default]
aws_access_key_id = ACCOUNT-B-KEY
aws_secret_access_key = ACCOUNT-B-SEC
aws_account_id = ACCOUNT-B-ID


[sharedaccount]
aws_access_key_id = ACCOUNT-A-KEY
aws_secret_access_key = ACCOUNT-A-SEC
aws_account_id = ACCOUNT-A-ID
```

Personally, I prefer to use different files is because the original `.awscred` file in the cluster profile won't be touched.

patrickdillon commented:
/pj-rehearse

patrickdillon commented on ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh:
```diff
@@ -8,6 +8,11 @@ set -o pipefail
 trap 'save_stack_events_to_artifacts' EXIT TERM INT
 
 export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred"
+if [[ ${ENABLE_SHARED_VPC} == "yes" ]]; then
+  CLUSTER_CREATOR_USER_ARN=$(aws sts get-caller-identity | jq -r '.Arn')
+  CLUSTER_CREATOR_AWS_ACCOUNT_NO=$(echo $CLUSTER_CREATOR_USER_ARN | awk -F ":" '{print $5}')
+  export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred_shared_account"
+fi
```
Based on [the rehearsal,](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/39946/rehearse-39946-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws/1668629045471875072) it looks like this change might not be working:

```
[36mINFO[0m[2023-06-13T14:41:33Z] Running step installer-rehearse-aws-aws-provision-vpc-shared. 
[36mINFO[0m[2023-06-13T14:41:47Z] Logs for container test in pod installer-rehearse-aws-aws-provision-vpc-shared: 
[36mINFO[0m[2023-06-13T14:41:47Z] 
Unable to locate credentials. You can configure credentials by running "aws configure".

Unable to locate credentials. You can configure credentials by running "aws configure".
{"component":"entrypoint","error":"wrapped process failed: exit status 253","file":"k8s.io/test-infra/prow/entrypoint/run.go:79","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2023-06-13T14:41:47Z"}
error: failed to execute wrapped command: exit status 253 
[36mINFO[0m[2023-06-13T14:41:47Z] Step installer-rehearse-aws-aws-provision-vpc-shared failed after 13s. 
[36mINFO[0m[2023-06-13T14:41:47Z] Step phase pre failed after 44s.    
```

patrickdillon commented on ci-operator/step-registry/ipi/conf/aws/oidc-creds-provision/ipi-conf-aws-oidc-creds-provision-commands.sh:
```diff
@@ -64,3 +68,12 @@ echo -e "\n"
+for f in openshift*.yaml
+do
+  [[ -e "$f" ]] || exit 1
+  role_arn=$(yq-go r ${f} 'stringData.credentials' | grep role_arn | awk '{print $3}')
```
looks like the use of `yq-go` here is causing the failure in the sts rehersal: 
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_release/39946/rehearse-39946-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-sts/1668629045413154816/build-log.txt

yunjiang29 replied:
fixed, removed `yq-go`

yunjiang29 commented on ci-operator/step-registry/aws/provision/route53/private-hosted-zone/aws-provision-route53-private-hosted-zone-commands.sh:
```diff
@@ -23,3 +28,76 @@ CHANGE_ID="$(echo "${HOSTED_ZONE_CREATION}" | jq -r '.ChangeInfo.Id' | awk -F /
+            "Action": [
+                "route53:GetHostedZone",
+                "route53:ChangeResourceRecordSets",
+                "route53:ChangeTagsForResource"
```
Yes, this is what we need to do next, to figure out the exact necessary permissions.  

As we added [`ResourceGroupsandTagEditorFullAccess` and `AmazonRoute53FullAccess`](https://github.com/openshift/release/pull/39946/files#diff-9101fa1d5a290147c4b88eab97114d2c33d80dfa851716589ba444208541f6e4R103-R105) in the code,  this policy is not necessary, but we need to remove _FullAccess_ eventually, so I kept this custom policy.

mtulio replied:
if you have IAM permissions, the access analyzer generation policy may be helpful:
- https://docs.aws.amazon.com/IAM/latest/UserGuide/what-is-access-analyzer.html#what-is-access-analyzer-policy-generation
- https://docs.aws.amazon.com/IAM/latest/UserGuide/access-analyzer-policy-generation.html#access-analyzer-policy-generation-howitworks

yunjiang29 replied:
@mtulio Great, that's very helpful. I'll take a look and test it with my local script, and update this part later (maybe in separate PR)

yunjiang29 commented:
/retest

yunjiang29 commented:
/pj-rehearse

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-p1-f14

yunjiang29 commented:
@patrickdillon I haven't got a chance to deep dive into this issue, but it looks like installer can not find subnets if AZ specified in install-config.yaml:
```
level=error msg=failed to fetch Master Machines: failed to load asset "Install Config": failed to create install config: [controlPlane.platform.aws.zones: Invalid value: []string{"us-east-1a", "us-east-1b", "us-east-1c"}: No subnets provided for zones [us-east-1b], compute[0].platform.aws.zones: Invalid value: []string{"us-east-1a", "us-east-1b", "us-east-1c"}: No subnets provided for zones [us-east-1b]]
```

see rehearse job https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/39946/rehearse-39946-periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-p1-f14/1668890235003473920

jianlinliu commented on ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:
```diff
@@ -924,6 +924,33 @@ tests:
+    env:
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      E2E_RUN_TAGS: '@amd64 and @aws-ipi and @network-ovnkubernetes and not @fips'
+      FEATURE_SET: TechPreviewNoUpgrade
```
If this job is a techpreview feature, need to add "techpreview" keyword in the job name.

yunjiang29 replied:
Open question: do we consider using `-tp` instead of `-techpreview` in the future? b/c it's more shorter.

jianlinliu replied:
yeah, you can use `tp` instead of `techpreview`, in someday, we can keep all the job consistent.

jianlinliu commented on ci-operator/step-registry/ipi/conf/aws/custom-vpc/ipi-conf-aws-custom-vpc-commands.sh:
```diff
@@ -18,11 +18,18 @@ fi
+yq-go m -x -i "${CONFIG}" "${CONFIG_SUBNETS}"
+
+
+if [[ ${ADD_ZONES} == "yes" ]]; then
```
What case lead us have to set it as a separate parameter?

yunjiang29 replied:
It's a temporary workaround for 
https://github.com/openshift/release/pull/39946#issuecomment-1590836744

jianlinliu replied:
So it is some bug to track?

yunjiang29 replied:
working on it to double check

yunjiang29 replied:
https://issues.redhat.com/browse/OCPBUGS-15058

yunjiang29 commented:
/pj-rehearse cucushift-installer-rehearse-aws-ipi-shared-vpc-phz-sts

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-p1-f14

jianlinliu commented on ci-operator/step-registry/aws/deprovision/route53/private-hosted-zone/aws-deprovision-route53-private-hosted-zone-commands.sh:
```diff
@@ -22,3 +27,14 @@ CHANGE_ID="$(echo "${HOSTED_ZONE}" | jq -r '.ChangeInfo.Id' | awk -F / '{printf
+
+
+if [[ ${ENABLE_SHARED_PHZ} == "yes" ]]; then
+  export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred_shared_account"
```
In `aws-ipi-shared-vpc-phz` case, the public dns records are still going to be created, which account would those dns be created in?

jianlinliu replied:
> DNS Records in private and public zone and cluster are created by Account B

Which account is the public zone located? The `.awscred` one?

yunjiang29 replied:
> Which account is the public zone located? The `.awscred` one?

Yes

jianlinliu commented:
@yunjiang29 one question, "cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts" v.s "cucushift-installer-rehearse-aws-ipi-shared-vpc-phz-sts", per my understanding, the public cluster install would cover more than the private one, right?  Seem like "cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts" is a bit redundant.

yunjiang29 commented:
> @yunjiang29 one question, "cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts" v.s "cucushift-installer-rehearse-aws-ipi-shared-vpc-phz-sts", per my understanding, the public cluster install would cover more than the private one, right? Seem like "cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts" is a bit redundant.

@jianlinliu Right, the only reason I created this job is that the private cluster will be supported in ROSA, I want to make sure it can also be covered in OCP. So remove it to reduce our profiles?

yunjiang29 commented:
/pj-rehearse

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14

jianlinliu commented:
> the only reason I created this job is that the private cluster will be supported in ROSA, I want to make sure it can also be covered in OCP. So remove it to reduce our profiles?

Yeah, I think we can remove it, and we can cover it in ROSA, or in other ways, maybe we can add the following (removing `cucushift-installer-rehearse-aws-ipi-shared-vpc-phz-sts`)?
```
cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts
cucushift-installer-rehearse-aws-ipi-shared-vpc-phz
```

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14

yunjiang29 commented:
/pj-rehearse

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-private-shared-vpc-phz-sts-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-shared-vpc-phz-tp-p1-f14

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-private-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-private-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

patrickdillon commented:
I think the failure during destroy [in the rehearsal](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/39946/rehearse-39946-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz/1669606963157143552/artifacts/installer-rehearse-aws-shared-vpc-phz/ipi-deprovision-deprovision/artifacts/.openshift_install.log): `level=info msg="Failed to remove tag kubernetes.io/cluster/ci-op-7jjgdglf-f06f3-7vw6x: shared;` 

Is due to the [tagging of the VPC in the tags-for-byo-vpc step](https://steps.ci.openshift.org/reference/aws-provision-tags-for-byo-vpc#line22). The installer is finding the cluster tag on the shared vpc but doesn't have permission to delete them, which I believe is expected. I don't think we actually need to pre-create any shared tags here do we? The installer should create the shared tags as needed. Also I believe tags only need to be applied to the VPC itself--not the subnets.

Let me know what you think.

patrickdillon commented:
Also, great work that this is getting through e2e testing successfully!

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz

yunjiang29 commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz-sts

yunjiang29 commented:
@patrickdillon thanks for debugging

> Is due to the [tagging of the VPC in the tags-for-byo-vpc step](https://steps.ci.openshift.org/reference/aws-provision-tags-for-byo-vpc#line22). The installer is finding the cluster tag on the shared vpc but doesn't have permission to delete them, which I believe is expected. 

Yes, I think so, I added `ec2:DeleteTags` permission to the [policy](https://github.com/openshift/release/pull/39946/files#diff-9101fa1d5a290147c4b88eab97114d2c33d80dfa851716589ba444208541f6e4R43-R52), and it works now, see this [rehearse job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/39946/rehearse-39946-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.14-installer-rehearse-aws-shared-vpc-phz/1670306022377394176).

But the interesting thing is, this error only happened on `us-east-1`,  the destroyer works well on `us-east-2` without `ec2:DeleteTags` permission, @patrickdillon  do you know more about this?


> I don't think we actually need to pre-create any shared tags here do we? The installer should create the shared tags as needed. Also I believe tags only need to be applied to the VPC itself--not the subnets.

Adding these tags is because [Bug 2075072](https://bugzilla.redhat.com/show_bug.cgi?id=2075072), we added some [new tags](https://issues.redhat.com/browse/OCPQE-10362?focusedId=20332409&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20332409):
* `kubernetes.io/cluster/<infra-id>: shared` for BYO VPC
* `kubernetes.io/role/internal-elb: <empty>` for private subnet

I think they are still being used by Ingress e2e test, and adding `ec2:DeleteTags` permission is an acceptable solution, it doesn't impact installer testing.
cc @lihongan

yunjiang29 commented:
/unhold

yunjiang29 commented:
/pj-rehearse ack

jianlinliu commented on ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:
```diff
@@ -925,6 +925,33 @@ tests:
     - ref: idp-htpasswd
     - ref: openshift-extended-test-longduration
     workflow: cucushift-installer-rehearse-aws-ipi-private
+- as: aws-ipi-private-shared-vpc-phz-sts-tp-p1-f14
```
How about naming it to `p2`.

yunjiang29 replied:
Updated

jianlinliu commented on ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/private/shared-vpc-phz/sts/deprovision/cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts-deprovision-chain.yaml:
```diff
@@ -0,0 +1,11 @@
+  as: cucushift-installer-rehearse-aws-ipi-private-shared-vpc-phz-sts-deprovision
+  steps:
+  - chain: cucushift-installer-rehearse-aws-ipi-private-deprovision
+  - ref: aws-deprovision-route53-private-hosted-zone
```
Here we are not using `cucushift-installer-rehearse-aws-ipi-byo-route53-deprovision` chain?

yunjiang29 replied:
> Here we are not using `cucushift-installer-rehearse-aws-ipi-byo-route53-deprovision` chain?

Right, `gather-installer-boostrap-logs-in-bastion` and `aws-deprovision-s3buckets` are required by private cluster.

jianlinliu replied:
ah, got it.

jianlinliu commented on ci-operator/step-registry/ipi/conf/aws/byo-route53/ipi-conf-aws-byo-route53-commands.sh:
```diff
@@ -26,3 +26,24 @@ platform:
+  fi
+  echo "hosted zone role arn: ${ROLE_ARN}"
+  CONFIG_ROUTE53_PRIVATE_HOSTEDZONE_ROLE="/tmp/install-config-route53-private-hosted-zone-role.yaml.patch"
+cat > "${CONFIG_ROUTE53_PRIVATE_HOSTEDZONE_ROLE}" << EOF
```
Missing the leading whitespaces?

yunjiang29 replied:
Updated

jianlinliu commented on ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:
```diff
@@ -1094,6 +1121,61 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-aws-ipi-sdn
+- as: aws-ipi-shared-vpc-phz-tp-p1-f14
```
Suggest to update "p1" to "p2", too.

yunjiang29 replied:
Updated

jianlinliu commented on ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:
```diff
@@ -1094,6 +1121,61 @@ tests:
+    test:
+    - chain: openshift-e2e-test-qe-destructive
+    workflow: cucushift-installer-rehearse-aws-ipi-shared-vpc-phz
+- as: aws-ipi-shared-vpc-phz-sts-fips-tp-p1-f14
```
Suggest to update "p1" to "p2", too.

yunjiang29 replied:
Setting all profiles for this feature to p2? I thought to keep one on p1.

yunjiang29 replied:
Updated

jianlinliu replied:
> Setting all profiles for this feature to p2? 

Two reasons:
1. We only selected server 4.next configurations as P1, tough P1 has different meaning in our test plan, but before a overall update, let us keep consistent.
2. These jobs are still TP, even follow the definition in our test plan doc, it also should be p2.

jianlinliu commented:
/lgtm

jianlinliu commented:
cc @liangxia to review

liangxia commented:
/approve
